storybook/src/docs/components/DataGrid/DataGrid.stories.tsx

```diff
@@ -202,6 +203,34 @@ storiesOf("stories/components/DataGrid", module)
             </Box>
         );
     })
+    .add("useDynamicGridVisibilityModel", () => {
+        const columns: GridColDef[] = [
+            {
+                field: "fullName",
+                headerName: "Full name",
+                flex: 1,
+                renderCell: ({ row }) => `${row.firstName} ${row.lastName}`,
+            },
+            {
+                field: "firstName",
+                headerName: "First name",
+                flex: 1,
+            },
+            {
+                field: "lastName",
+                headerName: "Last name",
+                flex: 1,
+            },
+        ];
+
+        const dynamicGridVisibilityProps = useDynamicGridVisibilityModel({
+            fullName: { defaultView: false, compactView: true },
+            firstName: { defaultView: true, compactView: false },
+            lastName: { defaultView: true, compactView: false },
+        });
```

Comment:
What do you say to this alternative API?

```tsx
const dynamicGridVisibilityProps = useDynamicGridVisibilityModel({
    initial: ["firstName", "lastName"],
    compact: ["fullName"],
});
```

Reply:
# 💯 

That would be a lot more understandable 👍🏼 

`initial` might not be optimal though, as there is not really an "Initial" view, more a "Compact" and "Non-Compact" view, or a "Mobile" and "Desktop" view. 

I wanted to avoid the words "Mobile" and "Desktop" as the "Compact" view could also make sense for Tablet sizes 🤷🏼‍♂️

Reply:
It's basically a brakpoint, why not use breakpoint naming?

Reply:
```ts
const responsiveGridVisibilityProps = useResponsiveGridVisibilityModel({
    xs: ["firstName", "lastName"],
    md: ["fullName"],
});
```

@nsams do you mean something like this? 
I think this would imply that other breakpoint values can also be set, which I don't think is what we want here. 
Having only `default` and `compact` as possible values makes it clear that there are exactly two views, otherwise it might be unclear, what happens e.g. in the `sm` breakpoint.

Reply:
> which I don't think is what we want here.

why only two views?

Reply:
According to UX we will (in rare cases) need a mobile/compact view in addition to the default view. 
Supporting more views seems unnecessary imo, just adds complexity.

Reply:
My thought was that instead of adding a new concept (default + compact) we could use the already existing and familiar concept of breakpoints. And yes - the consequence is that multiple have to be supported.

Reply:
Then we'd have to rethink the concept...

In the current implementation, columns are shown in both views by default, when set to be shown explicitly in one of the two views the column is simply hidden in the other view. 

If we just add more views (for each breakpoint) we'd have to configure the columns for every breakpoint individually, otherwise it's not really clear, at which breakpoint a column should be hidden again. 
Like this for example, with a lot more duplicate configuration:

```ts
const responsiveGridVisibilityProps = useResponsiveGridVisibilityModel({
    xs: ["firstName", "lastName"],
    sm: ["firstName", "lastName"],
    md: ["fullName"],
    lg: ["fullName"],
    xl: ["fullName"],
});
```

Reply:
how would you do it if this wasn't a DataGrid but two boxes?

Reply:
> how would you do it if this wasn't a DataGrid but two boxes?

I don't understand this question? 😅

Reply:
sorry.

Let's say we have 2 boxes (divs): one containing firstName+lastName and one containing fullName.

How would you typically add responsive styling to show the first >= md and the second <= sm?

Reply:
what about https://mui.com/material-ui/customization/breakpoints/#api?

Reply:
> while certain others should always be shown.

those will have no responsiveness settings.

Reply:
> what about https://mui.com/material-ui/customization/breakpoints/#api?

Yes, that's the MUI way of doing it. This is just a helper function for `@media min-width`. 

---

Ok, I think I was thinking about it way to complicated 😅 

We could probably do it in a similar way to how the media-query approach would work by taking the previous breakpoint's values by default if nothing is set for that breakpoint. 
Then we could still always show columns, that are not defined in any breakpoint. 
Like this:

```ts
const responsiveGridVisibilityProps = useResponsiveGridVisibilityModel({
    xs: ["firstName", "lastName"], // Also used for `sm` as `sm` is not defined
    md: ["fullName"], // Also used for `lg` and `xl` as they are not defined
});
```


Or would this seem weird then? Should we make all columns hidden by default, when a config for this is defined?
Then we would have to define every column shown, at least once, like this:

```ts
const responsiveGridVisibilityProps = useResponsiveGridVisibilityModel({
    xs: ["foo", "bar", "firstName", "lastName"], // Also used for `sm` as `sm` is not defined
    md: ["foo", "bar", "fullName"], // Also used for `lg` and `xl` as they are not defined
});
```